### PR TITLE
Extend DisjointSet with necessary features for expression-based root mapping

### PR DESCRIPTION
**Pull request by naoyam**
DisjointSet is extended in #423. This PR only extracts the extension as a standalone PR.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/disjoint_set.h [hunk 1]
```diff
@@ -17,7 +19,7 @@ namespace cuda {
 //! DisjointSet::areEqual(a,b) checks if a and b belong same class
 //!
 //! \note The template type T is assumed to be hashable
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/disjoint_set.h:21]: We can remove this line of comment now if we are requiring user to supply a hash function.

File: torch/csrc/jit/codegen/cuda/disjoint_set.h [hunk 2]
```diff
@@ -72,6 +74,64 @@ class DisjointSet {
...
   }
 
+  //! Queries if an element exists in this set
+  bool contains(T a) const {
+    return entry_map.count(a) > 0;
+  }
+
+  //! Just inserts an element as its own. It is considered not
+  //! equivalent with anything else.
+  bool insert(T a) {
```
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/disjoint_set.h]: Just curious, what would be a scenario where we want to create a singleton equivalence class ,when there is no equivalence to establish. Would this be for perf optimization and debugging or there's a semantical need for this?

File: torch/csrc/jit/codegen/cuda/disjoint_set.h (hunk 1 above)
Comment (shmsong) [inline torch/csrc/jit/codegen/cuda/disjoint_set.h]: We can remove this line of comment now if we are requiring user to supply a hash function.

File: torch/csrc/jit/codegen/cuda/disjoint_set.h (hunk 2 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/disjoint_set.h]: Thanks. This is actually not used anymore, so I'm going to remove it.

The reason I added that was for convenience. If `a` is not in the set, `areEquivalent(a, a)` returns false, and that was not the response for a particular use case I had. 

I also thought that maybe `areEquivalent` should return always true if the two arguments are equal with `operator==`, even though the arguments are not registered. But I'm not sure what would be the best. If things aren't in the set, maybe returning false is more appropriate.

Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/disjoint_set.h]: I see, makes sense. I was originally thinking of `areEquivalent` like "this set has registered that they are equivalent", but I guess in the case where inputs are just identical in some other sense (`==`) but unseen before it's hard to say. So please feel free to use the way most convenient for you and we just want to make sure to reflect that on the API and doxygen strings.

**Review by shmsong — APPROVED** [review]
Overall this is very cool extension to the disjoint set data class, and thanks for addressing the questions and comments. Looks good to me now.

**Pull request merged**